### ISSUE TITLE
Garbage Collection: Array contents can get corrupted by the automatic collector as of 2023.8

### ISSUE THREAD

**Issue by Alphish** [YoYoGames/GameMaker-Bugs#2253]
### Description

I set up a system to process another project's scripts and analyse its JSDocs, with the intent to use it for generating GitHub documentation. I also set up an array with warnings to display to know which parts of which scripts should be fixed. However, some indices of the warnings array get replaced with a random gibberish for reasons I can't fathom.

Somehow, it's always the same array and the index 2 (i.e. third item) that gets corrupted, although other indices may corrupted as well. The corruption happens a few frames after the warning has been originally added (and the frame is somewhat random), so maybe there's some asynchronicity at play (the scripts are processed over a few frames after some get_open_filename call).

You can see glimpses of warnings getting progressively corrupted in the attached output.txt file, in particular how one entry gets replaced with `_color` for some reason. The logs are made after processing each script, in ToolboxScript.complete function. However, the bug is so obscure and happens to a specific array, so I can't quite figure out how to make a sample project out of it. >.<

Screenshot of the corrupted messages (at indices 1 and 2):
![image](https://github.com/YoYoGames/GameMaker-Bugs/assets/7540940/eac3b08b-fae7-451f-857c-4995d2fb5643)

[ContactUsPackage.zip](https://github.com/YoYoGames/GameMaker-Bugs/files/13428040/ContactUsPackage.zip)

[Docs Generator Test.zip](https://github.com/YoYoGames/GameMaker-Bugs/files/13428042/Docs.Generator.Test.zip)

### Steps To Reproduce

1. Download Docs Generator Test ZIP, which includes both Community Toolbox project and Docs Generator project.
2. Open the Docs Generator project.
3. Run it, click "Load toolbox" and navigate to Community Toolbox.yyp location.
4. After a few frames of processing it should produce a bunch of warnings, some of which get replaced with gibberish like on the screenshot. The logs print the warnings array to pinpoint when the corruption happens - sometimes it's after processing Shaders script, and sometimes after Strings script, which makes me suspect it's something asynchronous.

### How reliably can you recreate this issue using your steps above?

Always

### Which version of GameMaker are you reporting this issue for?

2023.8.2 (Monthly)

### Which platform(s) are you seeing the problem on?

Windows

### Contact Us Package Attached?

- [X] I have attached my Contact Us Package

### Sample Project Added?

- [X] I have included a small sample project

**Comment by tinkerer-red** [comment]
Can confirm this is a bug as my [Multiprocessing](https://github.com/tinkerer-red/MultiProcessing/tree/main) library suffered from the same issue.

The exact code I'm using to produce my packets:
```gml
static send_struct = function(_socket, _struct) {
	var _buff = encode_struct(_struct);
	network_send_packet(_socket, _buff, buffer_tell(_buff));
	buffer_delete(_buff);
}
static encode_struct = function(_struct) {
	var _buff = buffer_create(1, buffer_grow, 1);
	var _json = json_stringify(_struct);
	
	buffer_seek(_buff, buffer_seek_start, 0);
	buffer_write(_buff, buffer_string, _json);
	
	return _buff;
}
static decode_struct = function(_buff){
	buffer_seek(_buff, buffer_seek_start, 0);
	var _json = buffer_read(_buff, buffer_string);
	
	try {
		var _r = json_parse(_json);
	}
	catch(error) {
		var _r = undefined;
	}
	
	return _r;
}
```
If you remove the try/catch then you will get an error every once in a while. where a struct containing an array will generate garbage data.

**Comment by Alphish** [comment]
An update: After I added `gc_enable(false)` at the start, I no longer experienced the weird corruptions. Of course, the problem now is that I don't have automatic garbage collection anymore, so a more proper solution is in order. ^^'

Also, I tested it on Beta Runtime v2023.1100.0.455 and the problem still happens (and it likewise gets fixed when `gc_enable(false)` is used).

**Comment by SilentPhil** [comment]
We too are now facing a similar problem during the development of Norland. It's just that periodically some arrays are corrupted and the game can either throw an error that will be caught by try-catch, or just quietly crash without an error (which is the worst thing). 
But we can't afford to turn off garbage collection - the game is very complex and generates a lot of data.

**Comment by yymrennie** [comment]
This was specifically an issue with array_map() in the sample project though a similar issue affects other array_*() functions that take callback functions as parameters. If the parameter passed to the callback function was modified inside it, it was possible for memory corruption to occur.

A fix has been put in for these cases in 2024.2.

**Comment by SilentPhil** [comment]
@yymrennie 
How can a callback function be modified from within? Can you show an example?
I would like to check our code and fix it before updating to a new version.

**Comment by Alphish** [comment]
@SilentPhil It referred to parameters being changed inside function, like so:
```gml
var _parsed_items = array_map(_file_lines, function(_line) {
    _line = string_trim(_line); // <- that's where the parameter value is changed
    // do other stuff with line
});
```
So in this case, you want to replace the reassignment with a new variable (e.g. `var _trimmed_line = string_trim(_line)`).

**Comment by SilentPhil** [comment]
@Alphish 
But in @tinkerer-red case, his Multiprocessing project doesn't use array_* functions with predicate, yet he reported silent crashes.

**Comment by Alphish** [comment]
@SilentPhil It might not be just a single problem. There also has been a fix for #2627 which turned out to be improper handling of `foo ??= new Obj()` construct by the compiler.

Either way, when the next Beta comes around it would be good to check if the problem is no longer in the Multiprocessing project - implying one of the GC fixes ended up addressing the underlying issue - or if it persists - suggesting there's still another GC odd case that has yet to be addressed.

What I can tell is that when I disabled GC and applied the workaround I described above (replacing argument variable reassignment with new local variable definition) I no longer got crashes on 2023.8. So the workaround was sufficient for my use-case, and presumably won't be necessary anymore once the fix rolls around with 2024.2 version.
